selfdrive/updated.py

```diff
@@ -1,57 +1,267 @@
+  # pathnames of regular files (no directories, symlinks, or special files).
+  inode_map = {}
+  os.chdir(search_dir)
+  for root, dirs, files in os.walk('.', topdown=True):
+    for file_name in files:
+      full_name = os.path.join(search_dir, root, file_name)
+      st = os.stat(full_name)
+      if S_ISREG(st[ST_MODE]):
+        inode_map.update({st[ST_INO]: full_name})
+  return inode_map
+
+def dup_ovfs_object(inode_map, source_obj, target_dir):
+  # Given a relative pathname to copy, and a new target root, duplicate the
+  # source object in the target root, using hardlinks for regular files.
+
+  st = os.lstat(source_obj)
+  target_obj = os.path.join(target_dir, source_obj)
+
+  if S_ISREG(st[ST_MODE]):
+    # Hardlink all regular files; ownership and permissions are inherited.
+    link(inode_map[st[ST_INO]], target_obj)
+  else:
+    # Recreate all directories and symlinks; copy ownership and permissions.
+    if S_ISDIR(st[ST_MODE]):
+      os.mkdir(os.path.join(FINALIZED, source_obj), S_IMODE(st[ST_MODE]))
+    elif S_ISLNK(st[ST_MODE]):
+      os.symlink(os.readlink(source_obj), target_obj)
+      os.chmod(target_obj, S_IMODE(st[ST_MODE]), follow_symlinks=False)
+    else:
+      # Ran into a FIFO, socket, etc. Should not happen in OP install dir.
+      # Ignore without copying for the time being, revisit if necessary.
+      pass
+    os.chown(target_obj, st[ST_UID], st[ST_GID], follow_symlinks=False)
+    os.utime(target_obj, (st[ST_ATIME], st[ST_MTIME]), follow_symlinks=False)
+
+def finalize_from_ovfs():
+  # Take the current OverlayFS merged view and finalize a copy outside of
+  # OverlayFS, ready to be swapped-in at BASEDIR.
+  #
+  # The "copy" is done with hardlinks, but since the OverlayFS merge looks
+  # like a different filesystem, and hardlinks can't cross filesystems, we
+  # have to link from a pathname in the upper or lower layers.
 
-    # Write latest release notes to param
+  cloudlog.info("creating overlay inode index")
+  inode_map = inodes_in_tree(BASEDIR)
+  inode_map.update(inodes_in_tree(OVERLAY_UPPER))
+
+  cloudlog.info("creating finalized version of the overlay")
+  shutil.rmtree(FINALIZED)
+  os.umask(0o077)
+  os.mkdir(FINALIZED)
+  os.chdir(OVERLAY_MERGED)
+  for root, dirs, files in os.walk('.', topdown=True):
+    for obj_name in dirs:
+      dup_ovfs_object(inode_map, os.path.join(root, obj_name), FINALIZED)
+    for obj_name in files:
+      dup_ovfs_object(inode_map, os.path.join(root, obj_name), FINALIZED)
+  os.chdir(BASEDIR)  # otherwise umount will fail later with "target is busy"
+
+def attempt_update():
+  upd_lock_fd = open(UPDATE_LOCK, 'r')
+  try:
+    fcntl.flock(upd_lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+  except OSError:
+    print("couldn't get update lock; is another updated running? going back to sleep...")
+    upd_lock_fd.close()
+    return
+
+  cloudlog.info("attempting git update inside staging overlay")
+  os.chdir(OVERLAY_MERGED)
```

Comment:
Can we do all these things without calling `os.chdir`? That changes some global state that I don't think is needed.

Reply:
There were a couple of design reasons I did so, which have mostly been discarded at this point. I'll go back and tweak the directory walk and object copy operations to work with absolute paths instead of relative, and get rid of the chdirs.